Non-insertable and non-updatable properties (#1968)
oasis-tcs/odata-specs: odata-protocol/11.4 Data Modification.md

request body. The service MUST fail if unable to persist all property
values specified in the request.

Non-insertable properties SHOULD be omitted from the request body.

> If they are provided, services MUST either ignore the value in the request body or fail the request if the provided values do not match the service-determined values.

Author: Non-breaking alternative:
```suggestion
If they are provided, services MUST ignore the value in the request body.
```

request body. The service MUST fail if unable to persist all property
values specified in the request.

Non-insertable properties SHOULD be omitted from the request body.

> If they are provided, services MUST either ignore the values in the request body or fail the request,
> for example if the provided values do not match the service-determined values.

Author: ```suggestion
If they are provided, services MUST either ignore the values in the request body or fail the request
if the provided values do not match the service-determined values.
```

Options:
1. check metadata and don't look at the value if it is "read-only"
2. check metadata, look at the value, compute default, compare, then reject if different

Non-insertable properties include (and are not limited to)

- dependent properties that are tied to non-key properties of the principal entity through a referential constraint [OData-CSDL, section 8.5](#ODataCSDL) (informally: "denormalized" properties),
- properties annotated with the term
  [`Core.Computed`](https://github.com/oasis-tcs/odata-vocabularies/blob/main/vocabularies/Org.OData.Core.V1.md#Computed), see [OData-VocCore](#ODataVocCore),
- properties listed as `NonInsertableProperties` of term [`Capabilities.InsertRestrictions`](https://github.com/oasis-tcs/odata-vocabularies/blob/main/vocabularies/Org.OData.Capabilities.V1.md#InsertRestrictions), see [OData-VocCap](#ODataVocCap),
- properties annotated with term
  [`Core.Permissions`](https://github.com/oasis-tcs/odata-vocabularies/blob/main/vocabularies/Org.OData.Core.V1.md#Permissions), see [OData-VocCore](#ODataVocCore), where the annotation value does not have the `Write` flag.

Services MUST return an error if the request body contains a value for a
property that in principle can be inserted and currently cannot be inserted by the

> authenticated user (i.e., given the state of the object or permissions of the user).

Author: ```suggestion
authenticated user (for example, given the state of the object or permissions of the user).
```

> Services MUST return an error if the request body contains a value for a
> property that in principle can be inserted and currently cannot be inserted by the
> authenticated user (i.e., given the state of the object or permissions of the user).
> 

Author: ```suggestion
Services MUST return an error if the request body contains a value for a property that in principle can be specified on insert but cannot currently be set to the specified value, for example, due to permissions or state of the object.
```

a referential constraint on the dependent entity updates the dependent
property.

Non-updatable properties SHOULD be omitted from the request body.

> If they are provided, services MUST either ignore the values in the request body or fail the request,
> for example if the provided values do not match the actual values known by the service.

Author: ```suggestion
If they are provided, services MUST either ignore the values in the request body or fail the request
if the provided values do not match the actual values known by the service.
```

Non-updatable properties include (and are not limited to)

- key properties,
- dependent properties that are tied to non-key properties of the principal entity through a referential constraint [OData-CSDL, section 8.5](#ODataCSDL) (informally: "denormalized" properties),
- properties annotated with the terms
  [`Core.Computed`](https://github.com/oasis-tcs/odata-vocabularies/blob/main/vocabularies/Org.OData.Core.V1.md#Computed) or [`Core.Immutable`](https://github.com/oasis-tcs/odata-vocabularies/blob/main/vocabularies/Org.OData.Core.V1.md#Immutable), see [OData-VocCore](#ODataVocCore),
- properties listed as `NonUpdatableProperties` of term [`Capabilities.UpdateRestrictions`](https://github.com/oasis-tcs/odata-vocabularies/blob/main/vocabularies/Org.OData.Capabilities.V1.md#UpdateRestrictions), see [OData-VocCap](#ODataVocCap),
- properties annotated with term
  [`Core.Permissions`](https://github.com/oasis-tcs/odata-vocabularies/blob/main/vocabularies/Org.OData.Core.V1.md#Permissions), see [OData-VocCore](#ODataVocCore), where the annotation value does not have the `Write` flag.

> Services MUST return an error if the request body contains a value for a
> property that in principle can be updated and currently cannot be updated by the
> authenticated user (i.e., given the state of the object or permissions of the user), and
> the specified value differs from the current value of the property.

Author: ```suggestion
Services MUST return an error if the request body contains a value for a
property that in principle can be specified on update but the request cannot currently be executed respecting the specified value, for example, due to permissions or state of the object.
```

Rationale: don't disappoint the client by silently ignoring provided information.
